src/Tools/AnalyzerRunner/Program.cs

```diff
@@ -0,0 +1,480 @@
+            }
+
+            ImmutableDictionary<ProjectId, ImmutableArray<Diagnostic>>.Builder projectDiagnosticBuilder = ImmutableDictionary.CreateBuilder<ProjectId, ImmutableArray<Diagnostic>>();
+            foreach (var task in projectDiagnosticTasks)
+            {
+                projectDiagnosticBuilder.Add(task.Key, await task.Value.ConfigureAwait(false));
+            }
+
+            return projectDiagnosticBuilder.ToImmutable();
+        }
+
+        /// <summary>
+        /// Returns a list of all analyzer diagnostics inside the specific project. This is an asynchronous operation.
+        /// </summary>
+        /// <param name="analyzers">The list of analyzers that should be used</param>
+        /// <param name="project">The project that should be analyzed</param>
+        /// <see langword="false"/> to use the behavior configured for the specified <paramref name="project"/>.</param>
+        /// <param name="cancellationToken">The cancellation token that the task will observe.</param>
+        /// <returns>A list of diagnostics inside the project</returns>
+        private static async Task<ImmutableArray<Diagnostic>> GetProjectAnalyzerDiagnosticsAsync(ImmutableArray<DiagnosticAnalyzer> analyzers, Project project, Options analyzerOptionsInternal, TelemetryCollector telemetryCollector, CancellationToken cancellationToken)
+        {
+            Utilities.WriteLine($"Running analyzers for {project.Name}", ConsoleColor.Gray);
+            var supportedDiagnosticsSpecificOptions = new Dictionary<string, ReportDiagnostic>();
+
+            // Report exceptions during the analysis process as errors
+            supportedDiagnosticsSpecificOptions.Add("AD0001", ReportDiagnostic.Error);
+
+            // update the project compilation options
+            var modifiedSpecificDiagnosticOptions = supportedDiagnosticsSpecificOptions.ToImmutableDictionary().SetItems(project.CompilationOptions.SpecificDiagnosticOptions);
+            var modifiedCompilationOptions = project.CompilationOptions.WithSpecificDiagnosticOptions(modifiedSpecificDiagnosticOptions);
+            var processedProject = project.WithCompilationOptions(modifiedCompilationOptions);
+
+            try
+            {
+                Compilation compilation = await processedProject.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                var newCompilation = compilation.RemoveAllSyntaxTrees().AddSyntaxTrees(compilation.SyntaxTrees.Select(t => t.WithRootAndOptions(t.GetRoot(), t.Options.WithFeatures(new[] { new KeyValuePair<string, string>("IOperation", "IOperation") }))));
+
+                CompilationWithAnalyzers compilationWithAnalyzers = newCompilation.WithAnalyzers(analyzers, new CompilationWithAnalyzersOptions(new AnalyzerOptions(ImmutableArray.Create<AdditionalText>()), null, analyzerOptionsInternal.ConcurrentAnalysis, logAnalyzerExecutionTime: true, reportSuppressedDiagnostics: analyzerOptionsInternal.ReportSuppressedDiagnostics));
```

Comment:
some analyzer require additionaltext, we need to think about how to support such analyzer.

Reply:
➡️ Follow-up issue.

Reply:
[https://github.com/dotnet/roslyn/issues/23108](https://github.com/dotnet/roslyn/issues/23108)

---
In reply to: [149863772](https://github.com/dotnet/roslyn/pull/23087#discussion_r149863772) [](ancestors = 149863772)